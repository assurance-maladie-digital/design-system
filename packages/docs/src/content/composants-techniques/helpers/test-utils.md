---
title: testUtils
description: Le helper `testUtils` contient un ensemble de fonctions utilitaires pour les tests unitaires.
---

<doc-tabs>

<doc-tab-item label="Utilisation">

<doc-alert type="info">

Le helper `testUtils` exporte également toutes les fonctions du package [Vue Test Utils](https://vue-test-utils.vuejs.org/).

</doc-alert>

### applyGlobalConfig

La fonction `applyGlobalConfig` permet d’appliquer la configuration globale pour les tests, elle est utilisée dans le fichier `setup.ts`.

```ts
import { applyGlobalConfig } from '@cnamts/vue-dot/src/helpers/testUtils';

applyGlobalConfig();
```

### installGlobalPlugins

La fonction `installGlobalPlugins` permet d’installer les plugins [Vue Input Facade](https://ronaldjerez.github.io/vue-input-facade/), [Vue Meta](https://vue-meta.nuxtjs.org/), [Vuex](https://vuex.vuejs.org/), [Vuetify](https://v2.vuetifyjs.com/) et Vue Dot pour les tests.

```ts
import { createLocalVue, installGlobalPlugins } from '@cnamts/vue-dot/src/helpers/testUtils';

const localVue = createLocalVue();

installGlobalPlugins(localVue);
```

### installRouter

La fonction `installRouter` permet d’installer [Vue Router](https://v3.router.vuejs.org/fr/) pour les tests.

```ts
import { createLocalVue, installRouter } from '@cnamts/vue-dot/src/helpers/testUtils';

const localVue = createLocalVue();

installRouter(localVue);
```

### createRouter

La fonction `createRouter` permet de créer une instance de [Vue Router](https://v3.router.vuejs.org/fr/) pour les tests.

```ts
import { createRouter } from '@cnamts/vue-dot/src/helpers/testUtils';

const routes = [];

const router = createRouter({
	routes
});
```

### createStore

La fonction `createStore` permet de créer une instance de [Vuex](https://vuex.vuejs.org/) pour les tests.

```ts
import { createStore } from '@cnamts/vue-dot/src/helpers/testUtils';

const store = createStore({
	state: {}
});
```

### createVuetifyInstance

La fonction `createVuetifyInstance` permet de créer une instance de [Vuetify](https://v2.vuetifyjs.com/) pour les tests.

```ts
import { createVuetifyInstance } from '@cnamts/vue-dot/src/helpers/testUtils';

const vuetify = createVuetifyInstance();
```

### mockVFormRef

La fonction `mockVFormRef` permet de mocker la référence d’un composant [VForm](https://v2.vuetifyjs.com/en/components/forms/) pour les tests.

```ts
import { mockVFormRef } from '@cnamts/vue-dot/src/helpers/testUtils';

const mocks = {
	$refs: {
		form: mockVFormRef(true)
	}
};
```

### mockTranslations

La fonction `mockTranslations` permet de mocker les traductions pour les tests.

```ts
import { mockTranslations } from '@cnamts/vue-dot/src/helpers/testUtils';

const mocks = {
	...mockTranslations<string | string[]>({
		'path.to.title': 'Titre',
		'path.to.items': [
			'Élément 1',
			'Élément 2',
			'Élément 3'
		]
	})
};
```

</doc-tab-item>

<doc-tab-item label="API">
<doc-api name="helpers/test-utils"></doc-api>
</doc-tab-item>

</doc-tabs>
