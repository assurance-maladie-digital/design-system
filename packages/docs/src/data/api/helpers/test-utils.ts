import { Api } from '~/types';

export const api: Api = {
	applyGlobalConfig: {
		returnValue: [
			{
				type: 'void',
			},
		],
	},
	installGlobalPlugins: {
		arguments: [
			{
				name: 'localVue',
				description: 'Local Vue instance',
				type: 'Vue',
				required: true,
			},
		],
		returnValue: [
			{
				type: 'void',
			},
		],
	},
	installRouter: {
		arguments: [
			{
				name: 'localVue',
				description: 'Local Vue instance',
				type: 'Vue',
				required: true,
			},
		],
		returnValue: [
			{
				type: 'void',
			},
		],
	},
	createRouter: {
		arguments: [
			{
				name: 'options',
				description: 'Router options',
				type: 'RouterOptions',
				required: false,
			},
		],
		returnValue: [
			{
				type: 'VueRouter',
			},
		],
	},
	createStore: {
		arguments: [
			{
				name: 'options',
				description: 'Store options',
				type: 'StoreOptions',
				required: false,
			},
		],
		returnValue: [
			{
				type: 'Store',
			},
		],
	},
	createVuetifyInstance: {
		arguments: [
			{
				name: 'preset',
				description: 'Vuetify preset',
				type: 'Partial<UserVuetifyPreset>',
				required: true,
			},
		],
		returnValue: [
			{
				type: 'Vuetify',
			},
		],
	},
	mockVFormRef: {
		arguments: [
			{
				name: 'isValid',
				description: 'Is valid',
				type: 'boolean',
				required: true,
			},
		],
		returnValue: [
			{
				type: 'void',
			},
		],
	},
	mockTranslations: {
		arguments: [
			{
				name: 'translations',
				description: 'Translations',
				type: 'IndexedObject',
				required: true,
			},
		],
		returnValue: [
			{
				type: 'void',
			},
		],
	},
};
