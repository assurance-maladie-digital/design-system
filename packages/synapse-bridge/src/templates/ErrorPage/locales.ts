export const locales = {
	errorCodeText: 'Code d\'erreur\xa0: ',
	supportIdMessage: 'Votre identifiant de support\xa0:',
	btnText: 'Retour à l’accueil'
}
