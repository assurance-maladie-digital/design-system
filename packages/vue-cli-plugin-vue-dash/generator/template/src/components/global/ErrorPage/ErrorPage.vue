<template>
	<PageCard
		width="680"
		card-class="px-5"
	>
		<span
			v-if="code"
			class="font-weight-bold primary--text code"
		>
			{{ code }}
		</span>

		<h2
			:class="{
				'text-h4': !code,
				'mb-4': !code,
				'text-h6': code
			}"
			class="mb-2 font-weight-bold"
		>
			{{ title }}
		</h2>

		<p>{{ message }}</p>

		<VLayout
			v-if="cta && btnRoute"
			class="mt-6"
		>
			<VSpacer />

			<VBtn
				:to="btnRoute"
				exact
				color="primary"
			>
				{{ cta }}
			</VBtn>
		</VLayout>
	</PageCard>
</template>

<script lang="ts">
	import Vue from 'vue';
	import Component from 'vue-class-component';

	const Props = Vue.extend({
		props: {
			/** The HTTP code to display (optional) */
			code: {
				type: String,
				default: undefined
			},
			/** The title of the ErrorPage (if no code is passed, it will be bigger) */
			title: {
				type: String,
				required: true
			},
			/** Displays a message in a paragraph to the user */
			message: {
				type: String,
				required: true
			},
			/** Display a link to 'home' route when a text is passed */
			cta: {
				type: String,
				default: undefined
			},
			/** The route of the CTA button, default to home page */
			btnRoute: {
				type: [Array, Object],
				default: () => ({ name: 'home' })
			}
		}
	});

	/** Used to display error pages */
	@Component
	export default class ErrorPage extends Props {}
</script>

<style lang="scss" scoped>
	.code {
		font-size: 6rem;
		font-family: serif;
	}
</style>
