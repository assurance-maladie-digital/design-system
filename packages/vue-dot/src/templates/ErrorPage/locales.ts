export const locales = {
	supportIdMessage: 'Votre identifiant de support\xa0:',
	btnText: 'Retour à l’accueil'
};
