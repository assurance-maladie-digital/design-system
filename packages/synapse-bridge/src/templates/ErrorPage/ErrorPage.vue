<script lang="ts">
	import type { PropType } from 'vue'
	import { defineComponent } from 'vue'
	import { RouteRecordRaw } from 'vue-router'
	import { locales } from './locales'

	import PageContainer from '@/elements/PageContainer'

	export default defineComponent({
		components: {
			PageContainer,
		},
		props: {
			pageTitle: {
				type: String,
				default: undefined,
			},
			message: {
				type: String,
				default: undefined,
			},
			code: {
				type: String,
				default: undefined,
			},
			btnText: {
				type: String,
				default: locales.btnText,
			},
			btnRoute: {
				type: [Array, Object, String] as PropType<RouteRecordRaw | string>,
				default: () => ({ name: 'home' }),
			},
			btnHref: {
				type: String,
				default: undefined,
			},
			noBtn: {
				type: Boolean,
				default: false,
			},
			codeErrorText: {
				type: String,
				default: locales.errorCodeText
			},
		},
		data() {
			return {
				locales,
				route: this.btnRoute as RouteRecordRaw | string | undefined,
			}
		},
		created() {
			if (this.btnHref) {
				this.route = undefined
			}
		},
		computed: {
			mobileVersion(): boolean {
				return this.$vuetify.display.name === 'xs' || this.$vuetify.display.name === 'sm'
			},
		},
	})
</script>

<template>
	<PageContainer size="m">
		<VCard
			:elevation="mobileVersion ? 0 : 2"
			class="pa-6 pa-sm-16"
		>
			<VRow class="max-width-none">
				<VCol
					:sm="$slots.illustration ? 6 : 12"
					cols="12"
					class="order-last order-sm-first text-center text-sm-left"
				>
					<div class="vd-code font-weight-thin text-primary mb-4">
						<span class="d-sr-only">{{ codeErrorText }}</span> {{ code }}
					</div>

					<h2 class="mb-2 font-weight-bold text-h5 mb-4">
						{{ pageTitle }}
					</h2>

					<p>
						{{ message }}
					</p>

					<slot name="additional-content" />

					<slot name="action">
						<VBtn
							v-if="!noBtn && btnText && (route || btnHref)"
							:to="route"
							:href="btnHref"
							color="primary"
							exact
							class="mt-6"
						>
							{{ btnText }}
						</VBtn>
					</slot>
				</VCol>

				<VCol
					v-if="$slots.illustration"
					cols="12"
					sm="6"
					class="d-flex align-center justify-center"
				>
					<slot name="illustration" />
				</VCol>
			</VRow>
		</VCard>
	</PageContainer>
</template>

<style lang="scss" scoped>
	.vd-code {
		font-size: 6rem;
		line-height: 6rem;
	}
</style>
