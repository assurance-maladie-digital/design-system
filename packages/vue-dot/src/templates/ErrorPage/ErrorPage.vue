<template>
	<PageContainer size="m">
		<VCard
			:elevation="mobileVersion ? 0 : 2"
			class="pa-6 pa-sm-16"
		>
			<VRow class="max-width-none">
				<VCol
					:sm="$slots.illustration ? 6 : 12"
					cols="12"
					class="order-last order-sm-first text-center text-sm-left"
				>
					<div
						aria-hidden="true"
						class="vd-code font-weight-thin primary--text mb-4"
					>
						{{ code }}
					</div>

					<h2 class="mb-2 font-weight-bold text-h5 mb-4">
						{{ pageTitle }}
					</h2>

					<p class="mb-0">
						{{ message }}
					</p>

					<slot name="additional-content" />

					<slot name="action">
						<VBtn
							v-if="!noBtn && btnText && btnRoute"
							:to="btnRoute"
							:href="btnHref"
							color="primary"
							exact
							class="mt-6"
						>
							{{ btnText }}
						</VBtn>
					</slot>
				</VCol>

				<VCol
					v-if="$slots.illustration"
					cols="12"
					sm="6"
					class="d-flex align-center justify-center"
				>
					<slot name="illustration" />
				</VCol>
			</VRow>
		</VCard>
	</PageContainer>
</template>

<script lang="ts">
	import Vue, { PropType } from 'vue';
	import Component, { mixins } from 'vue-class-component';

	import { RawLocation } from 'vue-router';

	import { locales } from './locales';

	const Props = Vue.extend({
		props: {
			pageTitle: {
				type: String,
				default: undefined
			},
			message: {
				type: String,
				default: undefined
			},
			code: {
				type: String,
				default: undefined
			},
			btnText: {
				type: String,
				default: locales.btnText
			},
			btnRoute: {
				type: [Array, Object, String] as PropType<RawLocation>,
				default: () => ({ name: 'home' })
			},
			btnHref: {
				type: String,
				default: undefined
			},
			noBtn: {
				type: Boolean,
				default: false
			}
		}
	});

	const MixinsDeclaration = mixins(Props);

	@Component
	export default class MaintenancePage extends MixinsDeclaration {
		locales = locales;

		get mobileVersion(): boolean {
			return this.$vuetify.breakpoint.xsOnly;
		}
	}
</script>

<style lang="scss" scoped>
	.vd-code {
		font-size: 6rem;
		line-height: 6rem;
	}
</style>
