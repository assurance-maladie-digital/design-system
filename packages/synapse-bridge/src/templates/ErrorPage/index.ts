import ErrorPage from './ErrorPage.vue';

export default ErrorPage;
