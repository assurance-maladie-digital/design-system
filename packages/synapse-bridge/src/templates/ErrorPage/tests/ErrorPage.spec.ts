import { describe, it, expect } from 'vitest'
import { mount } from '@vue/test-utils'
import { vuetify } from '@tests/unit/setup'

import ErrorPage from '../'

describe('ErrorPage', () => {
	it('renders correctly', () => {
		const wrapper = mount(ErrorPage, {
			global: {
				plugins: [vuetify],
			},
			propsData: {
				pageTitle: 'Error',
				message: 'Error message',
			},
		})

		expect(wrapper.html()).toMatchSnapshot()
	})

	it('renders correctly with undefined route', () => {
		const wrapper = mount(ErrorPage, {
			global: {
				plugins: [vuetify],
			},
			propsData: {
				pageTitle: 'Error',
				message: 'Error message',
				btnHref: 'https://google.com',
			},
		})

		expect(wrapper.html()).toMatchSnapshot()
	})

	it('renders correctly on mobile', () => {
		const wrapper = mount(ErrorPage, {
			global: {
				plugins: [vuetify],
			},
			propsData: {
				pageTitle: 'Error',
				message: 'Error message',
			},
		})

		wrapper.vm.$vuetify.display.name = 'sm'

		expect(wrapper.html()).toMatchSnapshot()
	})
})
