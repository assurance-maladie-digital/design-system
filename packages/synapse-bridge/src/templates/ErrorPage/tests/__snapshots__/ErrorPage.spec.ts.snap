// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ErrorPage > renders correctly 1`] = `
<div data-v-3e1b6721="" data-v-44bff9b6="" class="py-10 px-8 vd-page-container d-flex justify-center">
  <div data-v-3e1b6721="" class="v-sheet v-theme--light bg-transparent" style="width: 800px;">
    <div data-v-44bff9b6="" class="v-card v-theme--light v-card--density-default elevation-2 v-card--variant-elevated pa-6 pa-sm-16">
      <!---->
      <div class="v-card__loader">
        <div class="v-progress-linear v-theme--light v-locale--is-ltr" style="top: 0px; height: 0px; --v-progress-linear-height: 2px; left: 50%; transform: translateX(-50%);" role="progressbar" aria-hidden="true" aria-valuemin="0" aria-valuemax="100">
          <!---->
          <div class="v-progress-linear__background" style="width: 100%;"></div>
          <transition-stub name="fade-transition" appear="false" persisted="false" css="true">
            <div class="v-progress-linear__indeterminate">
              <div class="v-progress-linear__indeterminate long"></div>
              <div class="v-progress-linear__indeterminate short"></div>
            </div>
          </transition-stub>
          <!---->
        </div>
      </div>
      <!---->
      <!---->
      <div data-v-44bff9b6="" class="v-row max-width-none">
        <div data-v-44bff9b6="" class="v-col-sm-12 v-col-12 order-last order-sm-first text-center text-sm-left">
          <div data-v-44bff9b6="" class="vd-code font-weight-thin text-primary mb-4"><span data-v-44bff9b6="" class="d-sr-only">Code d'erreur&nbsp;: </span> </div>
          <h2 data-v-44bff9b6="" class="mb-2 font-weight-bold text-h5 mb-4">Error</h2>
          <p data-v-44bff9b6="">Error message</p><a data-v-44bff9b6="" class="v-btn v-btn--elevated v-theme--light bg-primary v-btn--density-default v-btn--size-default v-btn--variant-elevated mt-6"><span class="v-btn__overlay"></span><span class="v-btn__underlay"></span>
            <!----><span class="v-btn__content" data-no-activator="">Retour à l’accueil</span>
            <!---->
            <!---->
          </a>
        </div>
        <!--v-if-->
      </div>
      <!---->
      <!----><span class="v-card__underlay"></span>
    </div>
  </div>
</div>
`;

exports[`ErrorPage > renders correctly on mobile 1`] = `
<div data-v-3e1b6721="" data-v-44bff9b6="" class="py-10 px-8 vd-page-container d-flex justify-center">
  <div data-v-3e1b6721="" class="v-sheet v-theme--light bg-transparent" style="width: 800px;">
    <div data-v-44bff9b6="" class="v-card v-theme--light v-card--density-default elevation-2 v-card--variant-elevated pa-6 pa-sm-16">
      <!---->
      <div class="v-card__loader">
        <div class="v-progress-linear v-theme--light v-locale--is-ltr" style="top: 0px; height: 0px; --v-progress-linear-height: 2px; left: 50%; transform: translateX(-50%);" role="progressbar" aria-hidden="true" aria-valuemin="0" aria-valuemax="100">
          <!---->
          <div class="v-progress-linear__background" style="width: 100%;"></div>
          <transition-stub name="fade-transition" appear="false" persisted="false" css="true">
            <div class="v-progress-linear__indeterminate">
              <div class="v-progress-linear__indeterminate long"></div>
              <div class="v-progress-linear__indeterminate short"></div>
            </div>
          </transition-stub>
          <!---->
        </div>
      </div>
      <!---->
      <!---->
      <div data-v-44bff9b6="" class="v-row max-width-none">
        <div data-v-44bff9b6="" class="v-col-sm-12 v-col-12 order-last order-sm-first text-center text-sm-left">
          <div data-v-44bff9b6="" class="vd-code font-weight-thin text-primary mb-4"><span data-v-44bff9b6="" class="d-sr-only">Code d'erreur&nbsp;: </span> </div>
          <h2 data-v-44bff9b6="" class="mb-2 font-weight-bold text-h5 mb-4">Error</h2>
          <p data-v-44bff9b6="">Error message</p><a data-v-44bff9b6="" class="v-btn v-btn--elevated v-theme--light bg-primary v-btn--density-default v-btn--size-default v-btn--variant-elevated mt-6"><span class="v-btn__overlay"></span><span class="v-btn__underlay"></span>
            <!----><span class="v-btn__content" data-no-activator="">Retour à l’accueil</span>
            <!---->
            <!---->
          </a>
        </div>
        <!--v-if-->
      </div>
      <!---->
      <!----><span class="v-card__underlay"></span>
    </div>
  </div>
</div>
`;

exports[`ErrorPage > renders correctly with undefined route 1`] = `
<div data-v-3e1b6721="" data-v-44bff9b6="" class="py-10 px-8 vd-page-container d-flex justify-center">
  <div data-v-3e1b6721="" class="v-sheet v-theme--light bg-transparent" style="width: 800px;">
    <div data-v-44bff9b6="" class="v-card v-theme--light v-card--density-default elevation-2 v-card--variant-elevated pa-6 pa-sm-16">
      <!---->
      <div class="v-card__loader">
        <div class="v-progress-linear v-theme--light v-locale--is-ltr" style="top: 0px; height: 0px; --v-progress-linear-height: 2px; left: 50%; transform: translateX(-50%);" role="progressbar" aria-hidden="true" aria-valuemin="0" aria-valuemax="100">
          <!---->
          <div class="v-progress-linear__background" style="width: 100%;"></div>
          <transition-stub name="fade-transition" appear="false" persisted="false" css="true">
            <div class="v-progress-linear__indeterminate">
              <div class="v-progress-linear__indeterminate long"></div>
              <div class="v-progress-linear__indeterminate short"></div>
            </div>
          </transition-stub>
          <!---->
        </div>
      </div>
      <!---->
      <!---->
      <div data-v-44bff9b6="" class="v-row max-width-none">
        <div data-v-44bff9b6="" class="v-col-sm-12 v-col-12 order-last order-sm-first text-center text-sm-left">
          <div data-v-44bff9b6="" class="vd-code font-weight-thin text-primary mb-4"><span data-v-44bff9b6="" class="d-sr-only">Code d'erreur&nbsp;: </span> </div>
          <h2 data-v-44bff9b6="" class="mb-2 font-weight-bold text-h5 mb-4">Error</h2>
          <p data-v-44bff9b6="">Error message</p><a data-v-44bff9b6="" class="v-btn v-btn--elevated v-theme--light bg-primary v-btn--density-default v-btn--size-default v-btn--variant-elevated mt-6" href="https://google.com"><span class="v-btn__overlay"></span><span class="v-btn__underlay"></span>
            <!----><span class="v-btn__content" data-no-activator="">Retour à l’accueil</span>
            <!---->
            <!---->
          </a>
        </div>
        <!--v-if-->
      </div>
      <!---->
      <!----><span class="v-card__underlay"></span>
    </div>
  </div>
</div>
`;
